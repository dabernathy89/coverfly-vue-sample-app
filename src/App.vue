<script setup>
  import { ref } from 'vue';
  import { useStore } from './store';

  const store = useStore();

  const input = ref(0);
  const waiting = ref(false);
  const failed = ref(false);

  async function submit() {
    failed.value = false;
    waiting.value = true;
    try {
      await store.handleCalculation(input.value);
    } catch (err) {
      console.log({err});
      failed.value = true;
      input.value = 0;
    }
    waiting.value = false;
  }
</script>

<template>
  <form @submit.prevent="submit">
    <label for="calculator-input">Provide your input number here:</label><br>
    <input :disabled="waiting" id="calculator-input" v-model="input" type="number" />
    <input type="submit" :disabled="waiting" value="Submit">
  </form>

  <p v-if="failed">
    Please enter a number between 1 and 100.
  </p>

  <hr>

  <div v-for="response in store.history">
    {{ JSON.stringify(response) }}
  </div>
</template>

<style>
</style>
