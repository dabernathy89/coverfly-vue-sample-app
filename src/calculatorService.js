import db from "./database";
import getResult from './calculator';
const database = db.inputs;


function addToHistory(input, result, now) {
  if (typeof database[input] === 'undefined') {
    database[input] = {result: result, history: [now]};
    return;
  }

  database[input].history.push(now);
}

function validate(num) {
  if (!Number.isInteger(num) || num < 1 || num > 100) {
    throw 'Invalid input: number should be between 1 and 100.'
  }
}

function buildResponse(now, result, input) {
  const occurrences = database[input].history.length;
  const last_datetime = occurrences > 1
    ? database[input].history[occurrences - 2]
    : now;

  return {
    datetime: now,
    value: result,
    number: input,
    occurrences: occurrences,
    last_datetime: last_datetime,
  }
}

const calculator = {
  calculate(body) {
    const input = JSON.parse(body).input;
    validate(input);

    const now = new Date();
    const result = typeof database[input] !== 'undefined'
      ? database[input].result
      : getResult(input);

    addToHistory(input, result, now);

    const response = buildResponse(now, result, input);

    // Imitate HTTP request
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve(response);
      }, 100);
    });
  }
}

export default calculator;