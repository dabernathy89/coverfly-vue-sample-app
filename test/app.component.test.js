import { mount, flushPromises } from '@vue/test-utils';
import { vi } from 'vitest';
import App from '../src/App.vue';
import calculatorService from '../src/calculatorService';
import { createTestingPinia } from '@pinia/testing';

const timestamp = new Date(1656850173381);
vi.spyOn(calculatorService, 'calculate').mockResolvedValue({
  datetime: timestamp,
  value: 2640,
  number: 10,
  occurrences: 1,
  last_datetime: timestamp,
});

test('mount component', async () => {
  const wrapper = mount(App, {
    global: {
      plugins: [createTestingPinia({ stubActions: false })],
    },
  });

  expect(wrapper.text()).toContain('Provide your input number here');
  expect(wrapper.text()).toContain('Output');
  expect(wrapper.text()).toContain('Runs');
  expect(wrapper.html()).toMatchSnapshot();

  await wrapper.get('input[type=number]').text('10');
  await wrapper.get('form').trigger('submit');
  await flushPromises();

  expect(wrapper.text()).toContain('2640');
  expect(wrapper.html()).toMatchSnapshot();
});