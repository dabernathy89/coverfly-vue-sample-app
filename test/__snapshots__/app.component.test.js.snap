// Vitest Snapshot v1

exports[`mount component 1`] = `
"<form><label for=\\"calculator-input\\">Provide your input number here:</label><br><input id=\\"calculator-input\\" type=\\"number\\"><input type=\\"submit\\" value=\\"Submit\\"></form>
<!--v-if-->
<hr>
<table>
  <thead>
    <tr>
      <th>Input</th>
      <th>Output</th>
      <th># Runs</th>
      <th>Date</th>
      <th>Previous Run Date</th>
    </tr>
  </thead>
  <tbody></tbody>
</table>"
`;

exports[`mount component 2`] = `
"<form><label for=\\"calculator-input\\">Provide your input number here:</label><br><input id=\\"calculator-input\\" type=\\"number\\"><input type=\\"submit\\" value=\\"Submit\\"></form>
<!--v-if-->
<hr>
<table>
  <thead>
    <tr>
      <th>Input</th>
      <th>Output</th>
      <th># Runs</th>
      <th>Date</th>
      <th>Previous Run Date</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>10</td>
      <td>2640</td>
      <td>1</td>
      <td>Jul 3, 2022 7:09:33 AM</td>
      <td>Jul 3, 2022 7:09:33 AM</td>
    </tr>
  </tbody>
</table>"
`;
